File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.cpp
#include <string>
#include <vector>

using namespace std;

int dx[4] = { 1, 0, -1, 0 };
int dy[4] = { 0, 1, 0, -1 };

vector<int> solution(int rows, int columns, vector<vector<int>> queries) {
    vector<int> answer;

    vector<vector<int>> maps(rows, vector<int>(columns, 1));

    int num = 1;
    for (int i = 0; i < rows; ++i) {
        for (int j = 0; j < columns; ++j){
            maps[i][j] = num++;
        }
    }

    for (int i = 0; i < queries.size(); ++i){
        int topLeftY = queries[i][0] - 1;
        int topLeftX = queries[i][1] - 1;
        int bottomY = queries[i][2] - 1;
        int bottomX = queries[i][3] - 1;

        int minimum = 1000001;

        int curDir = 0;

        int curX = topLeftX;
        int curY = topLeftY;
        int saved = maps[topLeftY][topLeftX];
        int temp;
        while (true){
            int nx = curX + dx[curDir];
            int ny = curY + dy[curDir];

            if (nx > bottomX || nx < topLeftX || ny > bottomY || ny < topLeftY) {
                ++curDir;
                continue;
            }

            temp = maps[ny][nx];
            maps[ny][nx] = saved;
            saved = temp;
            curX = nx;
            curY = ny;

            minimum = min(minimum, saved);

            if (curX == topLeftX && curY == topLeftY)
                break;
        }


        answer.push_back(minimum);
    }
    return answer;
}

File: 프로그래머스/2/77485. 행렬 테두리 회전하기/행렬 테두리 회전하기.java
class Solution {
    public static final int[] dx = {1, 0, -1, 0}; // 우, 하, 좌, 상
    public static final int[] dy = {0, 1, 0, -1};

    public int[] solution(int rows, int columns, int[][] queries) {
        int[] answer = new int[queries.length];

        int[][] map = new int[rows][columns];
        int num = 1;
        for (int i = 0; i < rows; i++) {
            for (int j = 0; j < columns; j++) {
                map[i][j] = num++;
            }
        }

        for (int i = 0; i < queries.length; i++) {

            int y1 = queries[i][0] - 1;
            int x1 = queries[i][1] - 1;
            int y2 = queries[i][2] - 1;
            int x2 = queries[i][3] - 1;

            int curX = x1;
            int curY = y1;
            int curDir = 0;

            int cur = map[y1][x1];
            answer[i] = cur;
            int next = 0;
            int loop = (x2 - x1) * 2 + (y2 - y1) * 2;
            System.out.println(loop);
            while (loop-- > 0) {
                int nx = curX + dx[curDir];
                int ny = curY + dy[curDir];
                if (nx < x1 || nx > x2 || ny < y1 || ny > y2) {
                    curDir = (curDir + 1) % 4;
                    nx = curX + dx[curDir];
                    ny = curY + dy[curDir];
                }

                next = map[ny][nx];
                map[ny][nx] = cur;
                cur = next;
                answer[i] = Math.min(cur, answer[i]);
                curX = nx;
                curY = ny;
            }

        }
        return answer;
    }
}